### PR TITLE
Allow `/player spawn` spawn offline-mode player in online server; Code clean-up in `/player spawn` command

### PR DESCRIPTION
**Pull request by keuin**
First I did a clean-up in method `PlayerCommand.spawn` and `EntityPlayerMPFake.createFake`, making them cleaner and easy-to-read.

Then I added a carpet configuration named `fallBackToSpawningOfflinePlayers` which allows `/player spawn <playerName>` to spawn offline-mode players if there is no online-mode player has the name `<playerName>`, which allows you spawning fake players with more freedom on their names. If online-mode player with the specified ID exists, `/player spawn` will spawn an online-mode player, which behaves identical with previous versions.

### CODE REVIEW

**Comment by altrisi** [comment]
Opinion: If it works, I don't think the rule is necessary and Carpet should allow spawning those players by default. Haven't checked the code yet.

**Comment by keuin** [comment]
The switch is set to `false` and thus won't affect any feature by default.

**Comment by gnembon** [comment]
sorry but code refactors are super unfun with two parallel versions of carpet running for two minecraft versions.

also ssuggesting using more 'noob friendly' name for 'fallBackToSpawningOfflinePlayers'

File: src/main/java/carpet/CarpetSettings.java [hunk 1]
```diff
@@ -28,17 +28,7 @@
...
-import static carpet.settings.RuleCategory.CREATIVE;
-import static carpet.settings.RuleCategory.EXPERIMENTAL;
-import static carpet.settings.RuleCategory.FEATURE;
-import static carpet.settings.RuleCategory.OPTIMIZATION;
-import static carpet.settings.RuleCategory.SURVIVAL;
-import static carpet.settings.RuleCategory.TNT;
-import static carpet.settings.RuleCategory.DISPENSER;
-import static carpet.settings.RuleCategory.SCARPET;
-import static carpet.settings.RuleCategory.CLIENT;
+import static carpet.settings.RuleCategory.*;
```
Comment (gnembon) [inline src/main/java/carpet/CarpetSettings.java:31]: please no

**Comment by gnembon** [comment]
don't import stars
don't reorder imports
don't refactor just to make it clearer
 - these last two are super important to track history changes even in between snapshots where only one branch is really important. otherwise tracking it gets unnecessarily complicated.

File: src/main/java/carpet/CarpetSettings.java (hunk 1 above)
Comment (keuin) [reply src/main/java/carpet/CarpetSettings.java]: I expanded the wildcard imports and reverted the reordering, keeping intact with your main stream.

**Comment by keuin** [comment]
Code changes are minimized and non-essential code refactors are reverted. Sorry for my casual modifications.

The method `cantSpawn` is inlined to method `EntityPlayerMPFake.createFake` because `cantSpawn` takes duplicate pre-condition checking and offline fake players can't be spawned unless the checking is removed.

**Comment by keuin** [comment]
As for the switch naming, I struggled to make some:

- `spawnOfflineFakePlayerIfUnavailable`
- `spawnOfflinePlayerInOnlineModeIfUnavailable`
- `allowSpawningOfflineFakePlayerInOnlineMode`

I prefer the second one. It is a little bit longer, but it is clearer.

File: src/main/java/carpet/patches/EntityPlayerMPFake.java [hunk 2]
```diff
@@ -11,45 +11,87 @@
...
+        PlayerManager playerManager = server.getPlayerManager();
+
+        // if this player already logged in
+        if (playerManager.getPlayer(playerName) != null)
+        {
+            throw new FakePlayerSpawnException("r Player ", "rb " + playerName, "r  is already logged on");
+        }
+
+        // check player name length
+        // player with a name longer than 16 bytes will kick out all players
```
Comment (gnembon) [inline src/main/java/carpet/patches/EntityPlayerMPFake.java:52]: not true - works in SP up to 40 chars

File: src/main/java/carpet/patches/EntityPlayerMPFake.java [hunk 3]
```diff
@@ -11,45 +11,87 @@
...
+        // if the player with this playerName does not exist
+        if (gameprofile == null)
+        {
+            if (!fallBackToSpawningOfflinePlayer) {
+                throw new FakePlayerSpawnException("r Player " + playerName + " is either banned by Mojang, or " +
+                        "auth servers are down. Banned players can only be summoned in Singleplayer " +
+                        "and in servers in off-line mode.");
+            } else {
+                // spawn offline player
+                gameprofile = new GameProfile(PlayerEntity.getOfflinePlayerUuid(playerName), playerName);
```
Comment (gnembon) [inline src/main/java/carpet/patches/EntityPlayerMPFake.java:68]: isn't that THE only line that really needed to be added to have the mvp in this situation?

File: src/main/java/carpet/commands/PlayerCommand.java [hunk 4]
```diff
@@ -178,37 +176,6 @@ private static boolean cantReMove(CommandContext<ServerCommandSource> context)
...
-        PlayerEntity player = manager.getPlayer(playerName);
-        if (player != null)
-        {
-            Messenger.m(context.getSource(), "r Player ", "rb " + playerName, "r  is already logged on");
-            return true;
-        }
-        GameProfile profile = server.getUserCache().findByName(playerName);
-        if (profile == null)
-        {
-            Messenger.m(context.getSource(), "r Player "+playerName+" is either banned by Mojang, or auth servers are down. " +
```
Comment (gnembon) [inline src/main/java/carpet/commands/PlayerCommand.java:195]: conditionally allow it here, by the rule.

File: src/main/java/carpet/patches/EntityPlayerMPFake.java (hunk 3 above)
Comment (keuin) [reply src/main/java/carpet/patches/EntityPlayerMPFake.java:68]: No, it isn't.
In file `PlayerCommand.java`, method `cantSpawn` checks the specified fake player name:

```java
    private static boolean cantSpawn(CommandContext<ServerCommandSource> context)
    {
        String playerName = StringArgumentType.getString(context, "player");
        MinecraftServer server = context.getSource().getMinecraftServer();
        PlayerManager manager = server.getPlayerManager();
        PlayerEntity player = manager.getPlayer(playerName);
        if (player != null)
        {
            Messenger.m(context.getSource(), "r Player ", "rb " + playerName, "r  is already logged on");
            return true;
        }
        GameProfile profile = server.getUserCache().findByName(playerName);
        if (profile == null)
        {
            Messenger.m(context.getSource(), "r Player "+playerName+" is either banned by Mojang, or auth servers are down. " +
                    "Banned players can only be summoned in Singleplayer and in servers in off-line mode.");
            return true;
        }
// ...
```

And if the online-mode player with specified name does not exist, this method will return `true`, which makes `spawn` method in `PlayerCommand.java` returns immediately:

```java
    private static int spawn(CommandContext<ServerCommandSource> context) throws CommandSyntaxException
    {
        if (cantSpawn(context)) return 0;
// ...
```

If only the line you mentioned was added, that line will never be reached.

**Comment by gnembon** [comment]
regarding namng, some more general wording might be more useful and potentially may extend on other usecases, like `strictOnlineMode` albeit I am not set on that either.

I have a feeling that this change can just be done in two if statements, one in 'cantSpawn` and the other in profile creations in constructor, unless I am wrong.

File: src/main/java/carpet/patches/EntityPlayerMPFake.java (hunk 2 above)
Comment (keuin) [reply src/main/java/carpet/patches/EntityPlayerMPFake.java:52]: Well, I saw that another PR is focusing on this problem, I will remove it from this PR.

File: src/main/java/carpet/patches/EntityPlayerMPFake.java (hunk 3 above)
Comment (gnembon) [reply src/main/java/carpet/patches/EntityPlayerMPFake.java:63]: well, I added the other comment above later
Allow it there (by the new rule): https://github.com/gnembon/fabric-carpet/pull/872/files#r637356921
and make it happen here, right?

Comment (keuin) [reply src/main/java/carpet/patches/EntityPlayerMPFake.java:63]: Yeah I re-implemented this feature in that way.

**Comment by keuin** [comment]
This feature is implemented without touching any other logic. I've tested it and it works in SMP and SSP.
However, I suggest cleaning method `createFake` and `cantSpawn` because currently they are fragile.

**Comment by altrisi** [comment]
Still don't get why should this be under a rule, since it's not like it's changing vanilla behaviour (more than fake players themselves are changing it), just upgrading Carpet behaviour. But I guess it's up to you.

I think such `strictOnlineMode` will actually confuse players and make them think they can use it to toggle online mode on and off during runtime.

**Comment by keuin** [comment]
People might accidentally spawn offline mode fake players which are not expected. A switch will give them control of the result. If they have not read the changelog, users may be confused.
I am not much inspired about naming, feel free to change it.

**Comment by gnembon** [comment]
maybe name is not that important, but I agree with keuin - as an admin, you may not want to choose one behaviour or the other. If you are using a certain player on and on for specific task, and with poor connection or something you may by acccident spawn a different player. Making sure they are all validated could be a tool an admin may want to use. Open for suggestions for a better, yet still short and to the point rule name for it.

**Comment by DragonEggBedrockBreaking** [comment]
I have some ideas:
`allowOfflineFakePlayers`
`allowSpawningOfflinePlayers`
`allowOfflinePLayerFallback`

**Comment by gnembon** [comment]
shorter, to the point, I like them

File: src/main/java/carpet/CarpetSettings.java [hunk 5]
```diff
@@ -472,6 +472,9 @@ public String description() {
     @Rule(desc = "Enables /player command to control/spawn players", category = COMMAND)
     public static String commandPlayer = "ops";
 
+    @Rule(desc = "Spawn offline players in online mode if online-mode player with specified name does not exist", category = COMMAND)
+    public static boolean strictOnlineMode = true;
```
Comment (gnembon) [inline src/main/java/carpet/CarpetSettings.java:476]:
```suggestion
    public static boolean allowSpawningOfflinePlayers = true;
```

File: src/main/java/carpet/commands/PlayerCommand.java [hunk 6]
```diff
@@ -192,9 +192,14 @@ private static boolean cantSpawn(CommandContext<ServerCommandSource> context)
         GameProfile profile = server.getUserCache().findByName(playerName);
         if (profile == null)
         {
-            Messenger.m(context.getSource(), "r Player "+playerName+" is either banned by Mojang, or auth servers are down. " +
-                    "Banned players can only be summoned in Singleplayer and in servers in off-line mode.");
-            return true;
+            if (CarpetSettings.strictOnlineMode)
```
Comment (gnembon) [inline src/main/java/carpet/commands/PlayerCommand.java:195]:
```suggestion
            if (!CarpetSettings.allowSpawningOfflinePlayers)
```

File: src/main/java/carpet/patches/EntityPlayerMPFake.java [hunk 7]
```diff
@@ -45,7 +45,15 @@ public static EntityPlayerMPFake createFake(String username, MinecraftServer ser
         finally {
             UserCache.setUseRemote(server.isDedicated() && server.isOnlineMode());
         }
-        if (gameprofile == null) return null;
+        if (gameprofile == null)
+        {
+            if (CarpetSettings.strictOnlineMode)
```
Comment (gnembon) [inline src/main/java/carpet/patches/EntityPlayerMPFake.java:50]:
```suggestion
            if (!CarpetSettings.allowSpawningOfflinePlayers)
```

**Review by gnembon — APPROVED** [review]

**Comment by DragonEggBedrockBreaking** [comment]
> shorter, to the point, I like them

yay, I did something useful \o/

**Pull request merged**